src/avatar/Avatar.tsx

```diff
@@ -15,87 +28,169 @@ export type AvatarProps = {
-    return (child as JSX.Element).type === AvatarBadge;
-  });
-  const elements = _children.filter(child => {
-    return (child as JSX.Element).type !== AvatarBadge;
-  });
+export { useAvatarContext };
 
-  return (
-    <div aria-label={name} {...rest} className={avatarStyles}>
-      <AvatarImage
-        src={src}
-        name={name}
-        onError={onError}
-        fallback={elements.length > 0 ? elements : fallback}
-      />
-      {React.Children.map(badges, badge => {
-        if (React.isValidElement(badge)) {
-          return React.cloneElement(badge, { size: _size });
-        }
-      })}
-    </div>
-  );
-};
+type AvatarRenderProps = RenderProp<AvatarContext>;
+
+export type AvatarProps = Omit<BoxProps, "onError"> &
+  AvatarInitialProps &
+  AvatarRenderProps;
+
+export const Avatar = forwardRefWithAs<AvatarProps, HTMLDivElement, "div">(
+  (props, ref) => {
+    const {
+      name,
+      src,
+      size,
+      onError,
+      fallback,
+      getInitials = initials,
+      loading,
+      status,
+      position,
+      showBorder,
+      borderColor,
+      children,
+      className,
+      ...rest
+    } = props;
+
+    const group = useAvatarGroup();
+    const _size = size || group?.size || "md";
+    const _showBorder = showBorder || group?.showBorder || false;
+    const _borderColor = borderColor || group?.borderColor || false;
+    const theme = useTheme();
+    const avatarStyles = cx(
+      theme.avatar.base,
+      theme.avatar.size[_size],
+      _showBorder ? theme.avatar.border.width[_size] : "",
+      _showBorder && _borderColor ? _borderColor : theme.avatar.border.color,
+      className,
+    );
+
+    /**
+     * Use the image hook to only show the image when it has loaded
+     */
+    const imageStatus = useImage({ src, onError });
+    const hasLoaded = imageStatus === "loaded";
+
+    /**
+     * Fallback avatar applies under 2 conditions:
+     * - If `src` was passed and the image has not loaded or failed to load
+     * - If `src` wasn't passed
+     *
+     * In this case, we'll show either the name avatar or default avatar
+     */
+    const showFallback = !src || !hasLoaded;
+
+    const context: AvatarContext = React.useMemo(
+      () => ({
+        size: _size,
+        src,
+        name,
+        loading,
+        getInitials,
+        onError,
+        fallback,
+        status,
+        position,
+        showFallback,
+      }),
+      [
+        _size,
+        src,
+        name,
+        loading,
+        getInitials,
+        onError,
+        fallback,
+        status,
+        position,
+        showFallback,
+      ],
+    );
 
-export type AvatarBadgeProps = {
-  position?: "top-left" | "top-right" | "bottom-right" | "bottom-left";
-} & Pick<AvatarProps, "size">;
+    return (
+      <AvatarProvider value={context}>
+        <Box ref={ref} className={avatarStyles} {...rest}>
+          {children ? (
+            runIfFn(children, context)
```

Comment:
pass getInitial function to renderProp

`runIfFn(children, {...context, getInitials})`

This way users can do 
```tsx
    <Avatar name="Anurag hazra">
      {({ name, status, src, getInitials }) => <p>{getInitials!(name)}</p>}
    </Avatar>
```


related to: https://github.com/timelessco/renderlesskit-react-tailwind/pull/63/files#r575204267